### DISCUSSION TITLE
woocommerce/product-pricing-field should not be full-width

### DISCUSSION THREAD

**Discussion by leewillis77** [woocommerce/woocommerce#43134] [WooCommerce New Product Editor]
I'm currently looking at supporting the new product editor in some of our WooCommerce extensions. I've followed the articles linked from [here](https://developer.woo.com/2023/11/17/getting-to-know-woo-extensibility-in-the-new-product-editor/) and have fairly quickly managed to get fields added to the editor. 

However, I feel that either the UX, or the developer experience is compromised by the default behaviour of the woocommerce/product-pricing-field field type. If you add the field to the product editor then by default it appears as full-width. E.g.

![Screenshot 2023-12-27 at 15 56 29](https://github.com/woocommerce/woocommerce/assets/1097338/0a517534-dc54-4617-9bf1-1e14b45319d2)

Example code:
```php
add_action(
	'woocommerce_block_template_area_product-form_after_add_block_product-pricing-conditional-wrapper',
	function( $field ) {
		$parent_field = $field->get_parent();
		$parent_field->add_block(
				[
					'id'         => 'example-field-id',
					'order'      => $field->get_order() + 1,
					'blockName'  => 'woocommerce/product-pricing-field',
					'attributes' => [
						'property' => 'meta.prop',
						'label'    => 'Example field'
					],
				]
		);
	}
);
```
This seems less-than-ideal from a UX point of view as the actual price is displayed so far away from the label making it difficult to scan. Even the standard fields ('List Price' & 'Sale Price' just above) seem wider than they could ideally be, although the effect there is lessened as they're embedded in a 2-column layout. 

Obviously, it's possible to replicate the behaviour of those built in fields by embedding the price field into a two-column layout with an empty second column, but that seems hacky, and significantly complicates the code required to add the field. In order to create this layout:
![Screenshot 2023-12-27 at 16 04 03](https://github.com/woocommerce/woocommerce/assets/1097338/1b209e5e-ed49-4248-8b64-26c315968576)

the code now looks like:
```php
add_action(
	'woocommerce_block_template_area_product-form_after_add_block_product-pricing-conditional-wrapper',
	function( $field ) {
		$parent_field = $field->get_parent();
		$field_id = 'example-field-id';
		$pricing_columns         = $parent_field->add_block(
			array(
				'id'        => $field_id . '-columns',
				'blockName' => 'core/columns',
				'order'     => $field->get_order() + 1,
			)
		);
		$pricing_column_1        = $pricing_columns->add_block(
			array(
				'id'         => $field_id . '-column-1',
				'blockName'  => 'core/column',
				'order'      => 10,
				'attributes' => array(
					'templateLock' => 'all',
				),
			)
		);
		$pricing_columns->add_block(
			array(
				'id'         => $field_id . '-column-2',
				'blockName'  => 'core/column',
				'order'      => 20,
				'attributes' => array(
					'templateLock' => 'all',
				),
			)
		);
		$pricing_column_1->add_block(
				[
					'id'         => 'example-field-id',
					'order'      => 10,
					'blockName'  => 'woocommerce/product-pricing-field',
					'attributes' => [
						'property' => 'meta.prop',
						'label'    => 'Example field'
					],
				]
		);
	}
);
```

I'm not quite sure what the actual solution would be, but it seems like the woocommerce/product-pricing-field should be sized appropriately based on the likely input field size required for a better UX and to avoid developers having to control it manually by embedding into columns.

**Comment by mattsherman** [comment]
Hi @leewillis77 !

Thanks for sharing your thoughts on this.

I agree that the solution that you had to implement (using columns) is a bit cumbersome.

 I've made a note of this so that the team can look into simplifying this scenario in the future.

**Comment by louwie17** [comment]
@j111q we would appreciate your thoughts on this, as this would require some design input.
Here are some of the options:
- We can add a specific max width to some of our generic blocks like the pricing fields for example ( shown above ). Currently they always extend to the max width.
     - In relation to that would there be other fields/blocks that need this?
- We could add support for a max width property on the generic blocks themselves, or some kind of class support similar to the bootstrap method like ( `col-2`, `col-12`, etc... )

At the moment our preference would be the first option, but curious if this will heavily impact the design?

**Comment by mattsherman** [comment]
Another option would be to keep the current solution (using columns), but make the insertion of nested blocks easier. @joshuatf is exploring some options in that regard.

**Comment by j111q** [comment]
@louwie17 @mattsherman IMO the ideal scenario is where developers can choose whether a field is max-width or 50% width, depending on the content and type of their field. That is how we structure it in our design components for all fields. For e.g., if they're showing a dropdown that has a long option name and long help text, maybe they'd want it to be max-width so the option name and help text is shown in full. But if they're showing a dropdown with options that are short, then it's a better UX for the field to be at half width.

I'm not too sure which of @louwie17's two options would be best for that ideal scenario - I'm guessing option 2?
